src/task/future.rs

```diff
@@ -0,0 +1,393 @@
+
+    use std::cell::RefCell;
+    use std::rc::Rc;
+    use std::sync::mpsc;
+
+    struct MockLocal {
+        runner: Rc<RefCell<Runner<MockLocal>>>,
+        task_rx: mpsc::Receiver<TaskCell<MockRemote, ()>>,
+        remote: MockRemote,
+    }
+
+    impl MockLocal {
+        fn new() -> MockLocal {
+            let (task_tx, task_rx) = mpsc::sync_channel(10);
+            MockLocal {
+                runner: Rc::new(RefCell::new(Runner::default())),
+                task_rx,
+                remote: MockRemote { task_tx },
+            }
+        }
+
+        /// Run `Runner::handle` once.
+        fn handle_once(&mut self) {
+            if let Ok(task_cell) = self.task_rx.try_recv() {
+                let runner = self.runner.clone();
+                runner.borrow_mut().handle(self, task_cell);
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    struct MockRemote {
+        task_tx: mpsc::SyncSender<TaskCell<MockRemote, ()>>,
+    }
+
+    impl LocalSpawn for MockLocal {
+        type TaskCell = TaskCell<MockRemote, ()>;
+        type Remote = MockRemote;
+
+        fn spawn(&mut self, task_cell: TaskCell<MockRemote, ()>) {
+            self.remote.task_tx.send(task_cell).unwrap();
+        }
+
+        fn remote(&self) -> Self::Remote {
+            self.remote.clone()
+        }
+    }
+
+    impl RemoteSpawn for MockRemote {
+        type TaskCell = TaskCell<MockRemote, ()>;
+
+        fn spawn(&self, task_cell: TaskCell<MockRemote, ()>) {
+            self.task_tx.send(task_cell).unwrap();
+        }
+    }
+
+    struct WakeLater {
+        waker_tx: mpsc::SyncSender<Waker>,
+        first_poll: bool,
+    }
+
+    impl WakeLater {
+        fn new(waker_tx: mpsc::SyncSender<Waker>) -> WakeLater {
+            WakeLater {
+                waker_tx,
+                first_poll: true,
+            }
+        }
+    }
+
+    impl Future for WakeLater {
+        type Output = ();
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+            if self.first_poll {
+                self.first_poll = false;
+                self.waker_tx
+                    .send(cx.waker().clone())
+                    .expect("waker channel disconnected");
+                Poll::Pending
+            } else {
+                Poll::Ready(())
+            }
+        }
+    }
+
+    fn test_wake_impl(f: impl FnOnce(Waker)) {
+        let mut local = MockLocal::new();
+        let (res_tx, res_rx) = mpsc::channel();
+        let (waker_tx, waker_rx) = mpsc::sync_channel(10);
+
+        let fut = async move {
+            res_tx.send(1).unwrap();
+            WakeLater::new(waker_tx.clone()).await;
+            res_tx.send(2).unwrap();
+        };
+        local.spawn(TaskCell::new(fut, local.remote(), ()));
+
+        local.handle_once();
+        assert_eq!(res_rx.recv().unwrap(), 1);
+        assert!(res_rx.try_recv().is_err());
+
+        let waker = waker_rx.recv().unwrap();
+        f(waker);
```

Comment:
How to verify it's really woken?

Reply:
I verify it by the code below. If `2` is received, the future must have been woken. Do you think the verification is too implicit? Would it be better to record it's spawned in `MockRemote` and check?

Reply:
Get it.